# Import Libraries and Data

In [865]:
import pandas as pd, numpy as np

In [866]:
train_data = pd.read_csv("allhyperdata.csv")
test_data = pd.read_csv("allhypertest.csv")

# Analyze Data pt. 1

### shape

In [867]:
train_data.shape

(2800, 30)

In [868]:
test_data.shape

(972, 30)

### columns

In [869]:
train_data.columns

Index(['Age', 'Sex', 'On Thyroxine', 'Query on Thyroxine',
       'On Antithyroid Medication', 'Sick', 'Pregnant', 'Thyroid Surgery',
       'I131 Treatement', 'Query Hypothyriod', 'Query Hyperthyriod', 'Lithium',
       'Goitre', 'Tumor', 'Hypopituitary', 'Psych', 'TSH Measured', 'TSH',
       'T3 Measured', 'T3', 'TT4 Measured', 'TT4', 'T4U Measured', 'T4U',
       'FTI Measured', 'FTI', 'TBG Measured', 'TBG', 'Referral Source',
       'Result'],
      dtype='object')

In [870]:
test_data.columns

Index(['Age', 'Sex', 'On Thyroxine', 'Query on Thyroxine',
       'On Antithyroid Medication', 'Sick', 'Pregnant', 'Thyroid Surgery',
       'I131 Treatement', 'Query Hypothyriod', 'Query Hyperthyriod', 'Lithium',
       'Goitre', 'Tumor', 'Hypopituitary', 'Psych', 'TSH Measured', 'TSH',
       'T3 Measured', 'T3', 'TT4 Measured', 'TT4', 'T4U Measured', 'T4U',
       'FTI Measured', 'FTI', 'TBG Measured', 'TBG', 'Referral Source',
       'Result'],
      dtype='object')

### ensure column names match

In [871]:
train_data.columns == test_data.columns

array([ True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True])

### adjust head() to show 50 columns

In [872]:
pd.set_option('display.max_columns', 50)

### head()

In [873]:
train_data.head()

Age Sex On Thyroxine Query on Thyroxine On Antithyroid Medication Sick  \
0  41   F            f                  f                         f    f   
1  23   F            f                  f                         f    f   
2  46   M            f                  f                         f    f   
3  70   F            t                  f                         f    f   
4  70   F            f                  f                         f    f   

  Pregnant Thyroid Surgery I131 Treatement Query Hypothyriod  \
0        f               f               f                 f   
1        f               f               f                 f   
2        f               f               f                 f   
3        f               f               f                 f   
4        f               f               f                 f   

  Query Hyperthyriod Lithium Goitre Tumor Hypopituitary Psych TSH Measured  \
0                  f       f      f     f             f     f            t   
1                  f       f      f     f             f     f            t   
2                  f       f      f     f             f     f            t   
3                  f       f      f     f             f     f            t   
4                  f       f      f     f             f     f            t   

    TSH T3 Measured   T3 TT4 Measured  TT4 T4U Measured   T4U FTI Measured  \
0   1.3           t  2.5            t  125            t  1.14            t   
1   4.1           t    2            t  102            f     ?            f   
2  0.98           f    ?            t  109            t  0.91            t   
3  0.16           t  1.9            t  175            f     ?            f   
4  0.72           t  1.2            t   61            t  0.87            t   

   FTI TBG Measured TBG Referral Source          Result  
0  109            f   ?            SVHC  negative.|3733  
1    ?            f   ?           other  negative.|1442  
2  120            f   ?           other  negative.|2965  
3    ?            f   ?           other   negative.|806  
4   70            f   ?             SVI  negative.|2807

In [874]:
test_data.head()

Age Sex On Thyroxine Query on Thyroxine On Antithyroid Medication Sick  \
0   35   F            f                  f                         f    f   
1   63   M            f                  f                         f    f   
2   25   F            f                  f                         f    f   
3   53   F            f                  f                         f    f   
4   92   F            f                  f                         f    f   

  Pregnant Thyroid Surgery I131 Treatement Query Hypothyriod  \
0        f               f               f                 f   
1        f               f               f                 f   
2        f               f               f                 f   
3        f               f               f                 t   
4        f               f               f                 f   

  Query Hyperthyriod Lithium Goitre Tumor Hypopituitary Psych TSH Measured  \
0                  f       f      f     f             f     f            f   
1                  f       f      f     f             f     f            t   
2                  f       f      f     f             f     f            t   
3                  f       f      f     f             f     f            t   
4                  f       f      f     f             f     f            t   

    TSH T3 Measured   T3 TT4 Measured  TT4 T4U Measured   T4U FTI Measured  \
0     ?           f    ?            f    ?            f     ?            f   
1   3.5           t  2.5            t  108            t  0.96            t   
2   4.6           t  2.4            t   61            t  0.82            t   
3  0.25           t  2.1            t  145            t  1.03            t   
4   0.7           t  1.3            t  120            t  0.84            t   

   FTI TBG Measured TBG Referral Source          Result  
0    ?            f   ?           other   negative.|219  
1  113            f   ?             SVI  negative.|2059  
2   75            f   ?            SVHD   negative.|399  
3  141            f   ?           other  negative.|1911  
4  143            f   ?             SVI   negative.|487

### info()

In [875]:
train_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2800 entries, 0 to 2799
Data columns (total 30 columns):
 #   Column                     Non-Null Count  Dtype 
---  ------                     --------------  ----- 
 0   Age                        2800 non-null   object
 1   Sex                        2800 non-null   object
 2   On Thyroxine               2800 non-null   object
 3   Query on Thyroxine         2800 non-null   object
 4   On Antithyroid Medication  2800 non-null   object
 5   Sick                       2800 non-null   object
 6   Pregnant                   2800 non-null   object
 7   Thyroid Surgery            2800 non-null   object
 8   I131 Treatement            2800 non-null   object
 9   Query Hypothyriod          2800 non-null   object
 10  Query Hyperthyriod         2800 non-null   object
 11  Lithium                    2800 non-null   object
 12  Goitre                     2800 non-null   object
 13  Tumor                      2800 non-null   object
 14  Hypopitu

In [876]:
test_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 972 entries, 0 to 971
Data columns (total 30 columns):
 #   Column                     Non-Null Count  Dtype 
---  ------                     --------------  ----- 
 0   Age                        972 non-null    int64 
 1   Sex                        972 non-null    object
 2   On Thyroxine               972 non-null    object
 3   Query on Thyroxine         972 non-null    object
 4   On Antithyroid Medication  972 non-null    object
 5   Sick                       972 non-null    object
 6   Pregnant                   972 non-null    object
 7   Thyroid Surgery            972 non-null    object
 8   I131 Treatement            972 non-null    object
 9   Query Hypothyriod          972 non-null    object
 10  Query Hyperthyriod         972 non-null    object
 11  Lithium                    972 non-null    object
 12  Goitre                     972 non-null    object
 13  Tumor                      972 non-null    object
 14  Hypopituit

### value_counts()

In [877]:
for column in train_data.columns:
    print("\n", train_data[column].value_counts(), "\n")


 59     75
70     74
60     66
63     64
72     64
       ..
455     1
6       1
?       1
93      1
92      1
Name: Age, Length: 94, dtype: int64 


 F    1830
M     860
?     110
Name: Sex, dtype: int64 


 f    2470
t     330
Name: On Thyroxine, dtype: int64 


 f    2760
t      40
Name: Query on Thyroxine, dtype: int64 


 f    2766
t      34
Name: On Antithyroid Medication, dtype: int64 


 f    2690
t     110
Name: Sick, dtype: int64 


 f    2759
t      41
Name: Pregnant, dtype: int64 


 f    2761
t      39
Name: Thyroid Surgery, dtype: int64 


 f    2752
t      48
Name: I131 Treatement, dtype: int64 


 f    2637
t     163
Name: Query Hypothyriod, dtype: int64 


 f    2627
t     173
Name: Query Hyperthyriod, dtype: int64 


 f    2786
t      14
Name: Lithium, dtype: int64 


 f    2775
t      25
Name: Goitre, dtype: int64 


 f    2729
t      71
Name: Tumor, dtype: int64 


 f    2799
t       1
Name: Hypopituitary, dtype: int64 


 f    2665
t     135
Name: Psych, dtype: in

In [878]:
for column in test_data.columns:
    print("\n", test_data[column].value_counts(), "\n")


 58    28
62    26
60    25
65    23
56    23
      ..
11     2
12     2
93     1
92     1
7      1
Name: Age, Length: 81, dtype: int64 


 F    650
M    282
?     40
Name: Sex, dtype: int64 


 f    838
t    134
Name: On Thyroxine, dtype: int64 


 f    962
t     10
Name: Query on Thyroxine, dtype: int64 


 f    963
t      9
Name: On Antithyroid Medication, dtype: int64 


 f    935
t     37
Name: Sick, dtype: int64 


 f    960
t     12
Name: Pregnant, dtype: int64 


 f    958
t     14
Name: Thyroid Surgery, dtype: int64 


 f    961
t     11
Name: I131 Treatement, dtype: int64 


 f    901
t     71
Name: Query Hypothyriod, dtype: int64 


 f    908
t     64
Name: Query Hyperthyriod, dtype: int64 


 f    968
t      4
Name: Lithium, dtype: int64 


 f    963
t      9
Name: Goitre, dtype: int64 


 f    947
t     25
Name: Tumor, dtype: int64 


 f    972
Name: Hypopituitary, dtype: int64 


 f    923
t     49
Name: Psych, dtype: int64 


 t    887
f     85
Name: TSH Measured, dtype

# Process Data pt. 1

### referral source irrelevant -> drop "Referral Source" column

In [879]:
train_data.drop(columns=["Referral Source"], inplace=True)
train_data.shape

(2800, 29)

In [880]:
test_data.drop(columns=["Referral Source"], inplace=True)
test_data.shape

(972, 29)

### TBG never measures -> drop "TBG Measured" and "TBG" columns

In [881]:
train_data.drop(columns=["TBG Measured", "TBG"], inplace=True)
train_data.shape

(2800, 27)

In [882]:
test_data.drop(columns=["TBG Measured", "TBG"], inplace=True)
test_data.shape

(972, 27)

### boolean columns "TSH Measured", "T3 Measured", "TT4 Measured", "T4U Measured", "FTI Measured", are redundant becuase the following column has a value or NaN

In [883]:
train_data.drop(columns=["TSH Measured", "T3 Measured", "TT4 Measured", "T4U Measured", "FTI Measured"], inplace=True)
train_data.shape

(2800, 22)

In [884]:
test_data.drop(columns=["TSH Measured", "T3 Measured", "TT4 Measured", "T4U Measured", "FTI Measured"], inplace=True)
test_data.shape

(972, 22)

### result value has gibberish at the end of the class
- **assumption** -> every result value has the format "*class*.*irrelevant*"
- must check by stripping the end of each string from the "." onwards
- and check value_counts() afterwards to see if every value falls into one of the 5 classes \["hyperthyroid", "t3 toxic", "goitre", "secondary toxic", "negative"]

In [885]:
def cleanStr(s):
    return s[:s.index(".")]
train_data["Result"] = train_data["Result"].apply(cleanStr)
train_data["Result"].value_counts()

negative        2723
hyperthyroid      62
T3 toxic           8
goitre             7
Name: Result, dtype: int64

In [886]:
test_data["Result"] = test_data["Result"].apply(cleanStr)
test_data["Result"].value_counts()

negative           947
hyperthyroid        17
goitre               5
T3 toxic             2
secondary toxic      1
Name: Result, dtype: int64

### drop 1 instance of secondary toxic
- do not have resources to collect more data
- do not have enough domain knowledge to generate synthetic data
- do not have enough domain knowledge to group with another class

In [887]:
mask = test_data["Result"] == "secondary toxic"
index = test_data.loc[mask].index
test_data.drop(index, inplace=True)
test_data["Result"].value_counts()

negative        947
hyperthyroid     17
goitre            5
T3 toxic          2
Name: Result, dtype: int64

### change "?" values in continuous columns to NaN
- "?" in categorical data will be encoded as the "missing" column

In [888]:
columns = ["Age", "TSH", "T3", "TT4", "T4U", "FTI"]
for column in columns:
    print(column)
    print((train_data[column] == "?").sum())
    train_data[column].replace("?", np.nan, inplace=True)
    print((train_data[column] == "?").sum())

Age
1
0
TSH
284
0
T3
585
0
TT4
184
0
T4U
297
0
FTI
295
0


In [889]:
columns = ["Age", "TSH", "T3", "TT4", "T4U", "FTI"]
for column in columns:
    print(column)
    print((test_data[column] == "?").sum())
    test_data[column].replace("?", np.nan, inplace=True)
    print((test_data[column] == "?").sum())

Age
0
0
TSH
85
0
T3
184
0
TT4
47
0
T4U
90
0
FTI
90
0


### change continuous values to proper data type

In [890]:
columns = ["Age", "TSH", "T3", "TT4", "T4U", "FTI"]
for column in columns:
    train_data[column] = train_data[column].astype("float")
train_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2800 entries, 0 to 2799
Data columns (total 22 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Age                        2799 non-null   float64
 1   Sex                        2800 non-null   object 
 2   On Thyroxine               2800 non-null   object 
 3   Query on Thyroxine         2800 non-null   object 
 4   On Antithyroid Medication  2800 non-null   object 
 5   Sick                       2800 non-null   object 
 6   Pregnant                   2800 non-null   object 
 7   Thyroid Surgery            2800 non-null   object 
 8   I131 Treatement            2800 non-null   object 
 9   Query Hypothyriod          2800 non-null   object 
 10  Query Hyperthyriod         2800 non-null   object 
 11  Lithium                    2800 non-null   object 
 12  Goitre                     2800 non-null   object 
 13  Tumor                      2800 non-null   objec

In [891]:
columns = ["Age", "TSH", "T3", "TT4", "T4U", "FTI"]
for column in columns:
    test_data[column] = test_data[column].astype("float")
test_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 971 entries, 0 to 971
Data columns (total 22 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Age                        971 non-null    float64
 1   Sex                        971 non-null    object 
 2   On Thyroxine               971 non-null    object 
 3   Query on Thyroxine         971 non-null    object 
 4   On Antithyroid Medication  971 non-null    object 
 5   Sick                       971 non-null    object 
 6   Pregnant                   971 non-null    object 
 7   Thyroid Surgery            971 non-null    object 
 8   I131 Treatement            971 non-null    object 
 9   Query Hypothyriod          971 non-null    object 
 10  Query Hyperthyriod         971 non-null    object 
 11  Lithium                    971 non-null    object 
 12  Goitre                     971 non-null    object 
 13  Tumor                      971 non-null    object 

### one-hot encoding of categorical data
- columns with categorical data includes \['Sex', 'On Thyroxine', 'Query on Thyroxine', 'On Antithyroid Medication', 'Sick', 'Pregnant', 'Thyroid Surgery', 'I131 Treatement', 'Query Hypothyriod', 'Query Hyperthyriod', 'Lithium', 'Goitre', 'Tumor', 'Hypopituitary', 'Psych', 'Result']

In [892]:
columns = train_data.columns[1:16] # Sex - Psych
temp_df = pd.DataFrame() # necessary so we can take these middle columns and put them in the middle of train_data
# encode columns
for column in columns:
    encoded_df = pd.get_dummies(train_data[column], prefix=column)
    temp_df = pd.concat([temp_df, encoded_df], axis=1)
# new df
train_data = pd.concat([train_data["Age"], temp_df, train_data.loc[:,"TSH":]], axis=1) # do not encode result (most scikit learn models can handle label)
train_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2800 entries, 0 to 2799
Data columns (total 38 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   Age                          2799 non-null   float64
 1   Sex_?                        2800 non-null   uint8  
 2   Sex_F                        2800 non-null   uint8  
 3   Sex_M                        2800 non-null   uint8  
 4   On Thyroxine_f               2800 non-null   uint8  
 5   On Thyroxine_t               2800 non-null   uint8  
 6   Query on Thyroxine_f         2800 non-null   uint8  
 7   Query on Thyroxine_t         2800 non-null   uint8  
 8   On Antithyroid Medication_f  2800 non-null   uint8  
 9   On Antithyroid Medication_t  2800 non-null   uint8  
 10  Sick_f                       2800 non-null   uint8  
 11  Sick_t                       2800 non-null   uint8  
 12  Pregnant_f                   2800 non-null   uint8  
 13  Pregnant_t        

In [893]:
columns = test_data.columns[1:16]
temp_df = pd.DataFrame()
for column in columns:
    encoded_df = pd.get_dummies(test_data[column], prefix=column)
    temp_df = pd.concat([temp_df, encoded_df], axis=1)
# encoded_result_df = pd.get_dummies(test_data["Result"], prefix="Result")
# test_data = pd.concat([test_data["Age"], temp_df, test_data.loc[:,"TSH":"FTI"], encoded_result_df], axis=1)
test_data = pd.concat([test_data["Age"], temp_df, test_data.loc[:,"TSH":]], axis=1) # do not encode result
test_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 971 entries, 0 to 971
Data columns (total 37 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   Age                          971 non-null    float64
 1   Sex_?                        971 non-null    uint8  
 2   Sex_F                        971 non-null    uint8  
 3   Sex_M                        971 non-null    uint8  
 4   On Thyroxine_f               971 non-null    uint8  
 5   On Thyroxine_t               971 non-null    uint8  
 6   Query on Thyroxine_f         971 non-null    uint8  
 7   Query on Thyroxine_t         971 non-null    uint8  
 8   On Antithyroid Medication_f  971 non-null    uint8  
 9   On Antithyroid Medication_t  971 non-null    uint8  
 10  Sick_f                       971 non-null    uint8  
 11  Sick_t                       971 non-null    uint8  
 12  Pregnant_f                   971 non-null    uint8  
 13  Pregnant_t          

### reindex() to match columns and fill missing columns with 0

In [894]:
# Ensure columns in the test set match the columns in the training set
test_data = test_data.reindex(columns=train_data.columns, fill_value=0)
train_data.columns == test_data.columns
test_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 971 entries, 0 to 971
Data columns (total 38 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   Age                          971 non-null    float64
 1   Sex_?                        971 non-null    uint8  
 2   Sex_F                        971 non-null    uint8  
 3   Sex_M                        971 non-null    uint8  
 4   On Thyroxine_f               971 non-null    uint8  
 5   On Thyroxine_t               971 non-null    uint8  
 6   Query on Thyroxine_f         971 non-null    uint8  
 7   Query on Thyroxine_t         971 non-null    uint8  
 8   On Antithyroid Medication_f  971 non-null    uint8  
 9   On Antithyroid Medication_t  971 non-null    uint8  
 10  Sick_f                       971 non-null    uint8  
 11  Sick_t                       971 non-null    uint8  
 12  Pregnant_f                   971 non-null    uint8  
 13  Pregnant_t          

### handle missing data
- count rows with a missing value
- count rows with a missing value AND a negative result
- if we have majority of missing values we can drop the rows to reduce the negative result and somewhat rebalance the class?
- i should try multiple methods of handling missing values and see how it affects the model performance.

In [895]:
train_data.dropna(inplace=True)
# columns = ["Age", "TSH", "T3", "TT4", "T4U", "FTI"]
# for column in columns:
#     mean = train_data[column].mean()
#     train_data[column].fillna(mean, inplace=True)
train_data.isna().sum().sum()

0

In [896]:
test_data.dropna(inplace=True)
# columns = ["Age", "TSH", "T3", "TT4", "T4U", "FTI"]
# for column in columns:
#     mean = test_data[column].mean()
#     test_data[column].fillna(mean, inplace=True)
test_data.isna().sum().sum()

0

# Analyze Data pt. 2

### describe() for continuous value columns

In [897]:
columns = ["Age", "TSH", "T3", "TT4", "T4U", "FTI"]
for column in columns:
    print(train_data[column].describe(), "\n")

count    2028.000000
mean       53.069527
std        20.837369
min         1.000000
25%        37.000000
50%        55.000000
75%        69.000000
max       455.000000
Name: Age, dtype: float64 

count    2028.000000
mean        4.555535
std        20.860154
min         0.005000
25%         0.450000
50%         1.300000
75%         2.500000
max       472.000000
Name: TSH, dtype: float64 

count    2028.000000
mean        2.018146
std         0.822898
min         0.050000
25%         1.600000
50%         2.000000
75%         2.400000
max        10.600000
Name: T3, dtype: float64 

count    2028.000000
mean      108.710848
std        35.445779
min         2.000000
25%        88.000000
50%       104.000000
75%       124.000000
max       430.000000
Name: TT4, dtype: float64 

count    2028.000000
mean        0.998163
std         0.198381
min         0.310000
25%         0.880000
50%         0.980000
75%         1.090000
max         2.120000
Name: T4U, dtype: float64 

count    2028.000000


In [898]:
columns = ["Age", "TSH", "T3", "TT4", "T4U", "FTI"]
for column in columns:
    print(test_data[column].describe(), "\n")

count    722.000000
mean      52.400277
std       18.943134
min        7.000000
25%       36.000000
50%       55.000000
75%       68.000000
max       93.000000
Name: Age, dtype: float64 

count    722.000000
mean       6.712888
std       34.448055
min        0.005000
25%        0.500000
50%        1.300000
75%        2.800000
max      530.000000
Name: TSH, dtype: float64 

count    722.000000
mean       1.957950
std        0.811188
min        0.200000
25%        1.500000
50%        1.950000
75%        2.300000
max        8.500000
Name: T3, dtype: float64 

count    722.000000
mean     105.742244
std       35.090583
min        2.900000
25%       87.000000
50%      103.000000
75%      123.000000
max      372.000000
Name: TT4, dtype: float64 

count    722.000000
mean       0.987230
std        0.190557
min        0.250000
25%        0.870000
50%        0.970000
75%        1.070000
max        2.010000
Name: T4U, dtype: float64 

count    722.000000
mean     108.260111
std       32.362819
m

# Process Data pt. 2

### max age val of 455 in train_data is invalid -> replace it with the mean

In [899]:
mask = train_data["Age"] > 125
train_data.loc[mask, "Age"] = train_data["Age"].mean()
train_data["Age"].describe()

count    2028.000000
mean       52.871336
std        18.827069
min         1.000000
25%        37.000000
50%        55.000000
75%        69.000000
max        94.000000
Name: Age, dtype: float64

# Analyze Data pt. 3

In [900]:
train_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2028 entries, 0 to 2799
Data columns (total 38 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   Age                          2028 non-null   float64
 1   Sex_?                        2028 non-null   uint8  
 2   Sex_F                        2028 non-null   uint8  
 3   Sex_M                        2028 non-null   uint8  
 4   On Thyroxine_f               2028 non-null   uint8  
 5   On Thyroxine_t               2028 non-null   uint8  
 6   Query on Thyroxine_f         2028 non-null   uint8  
 7   Query on Thyroxine_t         2028 non-null   uint8  
 8   On Antithyroid Medication_f  2028 non-null   uint8  
 9   On Antithyroid Medication_t  2028 non-null   uint8  
 10  Sick_f                       2028 non-null   uint8  
 11  Sick_t                       2028 non-null   uint8  
 12  Pregnant_f                   2028 non-null   uint8  
 13  Pregnant_t        

In [901]:
test_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 722 entries, 1 to 971
Data columns (total 38 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   Age                          722 non-null    float64
 1   Sex_?                        722 non-null    uint8  
 2   Sex_F                        722 non-null    uint8  
 3   Sex_M                        722 non-null    uint8  
 4   On Thyroxine_f               722 non-null    uint8  
 5   On Thyroxine_t               722 non-null    uint8  
 6   Query on Thyroxine_f         722 non-null    uint8  
 7   Query on Thyroxine_t         722 non-null    uint8  
 8   On Antithyroid Medication_f  722 non-null    uint8  
 9   On Antithyroid Medication_t  722 non-null    uint8  
 10  Sick_f                       722 non-null    uint8  
 11  Sick_t                       722 non-null    uint8  
 12  Pregnant_f                   722 non-null    uint8  
 13  Pregnant_t          

# Train Models & Test Models

### split the data

In [902]:
X_train = train_data.iloc[:,:37]
y_train = train_data.iloc[:,37]
X_test = test_data.iloc[:,:37]
y_test = test_data.iloc[:,37]
y_train

0       negative
4       negative
7       negative
8       negative
9       negative
          ...   
2788    negative
2790    negative
2793    negative
2794    negative
2799    negative
Name: Result, Length: 2028, dtype: object

### train & test models

In [903]:
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report

# Assuming you have a dataset with features X and labels y
# X should be a 2D array (or DataFrame), and y should be a 1D array or pandas Series

# Create a RandomForestClassifier
rf_classifier = RandomForestClassifier(n_estimators=100, random_state=42)

# Train the classifier on the training set
rf_classifier.fit(X_train, y_train)

# Make predictions on the test set
y_pred = rf_classifier.predict(X_test)

# Evaluate the model
accuracy = accuracy_score(y_test, y_pred)
print(f'Accuracy: {accuracy:.2f}')

# Print classification report
print('\nClassification Report:')
print(classification_report(y_test, y_pred))

Accuracy: 0.99

Classification Report:
              precision    recall  f1-score   support

    T3 toxic       0.00      0.00      0.00         2
      goitre       1.00      1.00      1.00         4
hyperthyroid       0.83      0.50      0.62        10
    negative       0.99      1.00      0.99       706

    accuracy                           0.99       722
   macro avg       0.71      0.62      0.65       722
weighted avg       0.99      0.99      0.99       722



C:\Users\conno\AppData\Local\Programs\Python\Python310\lib\site-packages\sklearn\metrics\_classification.py:1471: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\conno\AppData\Local\Programs\Python\Python310\lib\site-packages\sklearn\metrics\_classification.py:1471: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\conno\AppData\Local\Programs\Python\Python310\lib\site-packages\sklearn\metrics\_classification.py:1471: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average,

In [904]:
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report

# Assuming your target variable is 'y' with categorical labels
# X should be a 2D array (or DataFrame), and y should be a 1D array or pandas Series

# Create a RandomForestClassifier
rf_classifier = RandomForestClassifier(class_weight="balanced")

# Define the hyperparameter grid to search
param_grid = {
    'n_estimators': [50, 100, 200],
    'max_depth': [None, 10, 20],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4],
}

# Create the GridSearchCV object
grid_search = GridSearchCV(rf_classifier, param_grid, cv=5, scoring='accuracy')

# Fit the model to the training data
grid_search.fit(X_train, y_train)

# Get the best parameters and the best model
best_params = grid_search.best_params_
best_model = grid_search.best_estimator_

# Make predictions on the test set using the best model
y_pred = best_model.predict(X_test)

# Evaluate the best model
accuracy = accuracy_score(y_test, y_pred)
print(f'Best Model Accuracy: {accuracy:.2f}')

# Print best parameters
print('\nBest Hyperparameters:')
print(best_params)

# Print classification report
print('\nClassification Report:')
print(classification_report(y_test, y_pred))

Best Model Accuracy: 0.99

Best Hyperparameters:
{'max_depth': 10, 'min_samples_leaf': 2, 'min_samples_split': 10, 'n_estimators': 100}

Classification Report:
              precision    recall  f1-score   support

    T3 toxic       0.00      0.00      0.00         2
      goitre       1.00      1.00      1.00         4
hyperthyroid       0.78      0.70      0.74        10
    negative       0.99      1.00      1.00       706

    accuracy                           0.99       722
   macro avg       0.69      0.67      0.68       722
weighted avg       0.99      0.99      0.99       722



C:\Users\conno\AppData\Local\Programs\Python\Python310\lib\site-packages\sklearn\metrics\_classification.py:1471: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\conno\AppData\Local\Programs\Python\Python310\lib\site-packages\sklearn\metrics\_classification.py:1471: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\conno\AppData\Local\Programs\Python\Python310\lib\site-packages\sklearn\metrics\_classification.py:1471: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average,

In [905]:
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, classification_report

# Create a GradientBoostingClassifier
gb_classifier = GradientBoostingClassifier(n_estimators=100, random_state=42)

# Train the classifier on the training set
gb_classifier.fit(X_train, y_train)

# Make predictions on the test set
y_pred = gb_classifier.predict(X_test)

# Evaluate the model
accuracy = accuracy_score(y_test, y_pred)
print(f'Accuracy: {accuracy:.2f}')

# Print classification report
print('\nClassification Report:')
print(classification_report(y_test, y_pred))

Accuracy: 0.99

Classification Report:
              precision    recall  f1-score   support

    T3 toxic       0.50      0.50      0.50         2
      goitre       1.00      0.75      0.86         4
hyperthyroid       0.78      0.70      0.74        10
    negative       0.99      1.00      0.99       706

    accuracy                           0.99       722
   macro avg       0.82      0.74      0.77       722
weighted avg       0.99      0.99      0.99       722



In [914]:
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.metrics import accuracy_score, classification_report

# Create a GradientBoostingClassifier
gb_classifier = GradientBoostingClassifier(random_state=42)

# Define the hyperparameter grid to search
param_grid = {
    'n_estimators': [90, 100, 110],
    'learning_rate': [0.08, 0.1, 0.12],
    'max_depth': [2, 3, 4]
}

# Create a GridSearchCV object
grid_search = GridSearchCV(estimator=gb_classifier, param_grid=param_grid, cv=5, scoring='accuracy', n_jobs=-1)

# Perform grid search on the training set
grid_search.fit(X_train, y_train)

# Get the best hyperparameters
best_params = grid_search.best_params_
print(f"Best Hyperparameters: {best_params}")

# Make predictions on the test set using the best model
best_model = grid_search.best_estimator_
y_pred = best_model.predict(X_test)

# Evaluate the best model
accuracy = accuracy_score(y_test, y_pred)
print(f'Accuracy: {accuracy:.2f}')

# Print classification report
print('\nClassification Report:')
print(classification_report(y_test, y_pred))

Best Hyperparameters: {'learning_rate': 0.08, 'max_depth': 4, 'n_estimators': 90}
Accuracy: 0.99

Classification Report:
              precision    recall  f1-score   support

    T3 toxic       0.25      0.50      0.33         2
      goitre       1.00      0.75      0.86         4
hyperthyroid       0.78      0.70      0.74        10
    negative       0.99      0.99      0.99       706

    accuracy                           0.99       722
   macro avg       0.76      0.74      0.73       722
weighted avg       0.99      0.99      0.99       722



In [907]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report
from imblearn.over_sampling import SMOTE

# Create a SMOTE object
smote = SMOTE(random_state=42)

# Apply SMOTE to the training set
X_train_resampled, y_train_resampled = smote.fit_resample(X_train, y_train)

# Create and train a RandomForestClassifier on the resampled training set
rf_classifier = RandomForestClassifier(n_estimators=100, random_state=42)
rf_classifier.fit(X_train_resampled, y_train_resampled)

# Make predictions on the test set
y_pred = rf_classifier.predict(X_test)

# Evaluate the model
accuracy = accuracy_score(y_test, y_pred)
print(f'Accuracy: {accuracy:.2f}')

# Print classification report
print('\nClassification Report:')
print(classification_report(y_test, y_pred))

Accuracy: 0.99

Classification Report:
              precision    recall  f1-score   support

    T3 toxic       0.00      0.00      0.00         2
      goitre       1.00      0.75      0.86         4
hyperthyroid       0.90      0.90      0.90        10
    negative       0.99      1.00      1.00       706

    accuracy                           0.99       722
   macro avg       0.72      0.66      0.69       722
weighted avg       0.99      0.99      0.99       722



C:\Users\conno\AppData\Local\Programs\Python\Python310\lib\site-packages\sklearn\metrics\_classification.py:1471: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\conno\AppData\Local\Programs\Python\Python310\lib\site-packages\sklearn\metrics\_classification.py:1471: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\conno\AppData\Local\Programs\Python\Python310\lib\site-packages\sklearn\metrics\_classification.py:1471: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average,

In [908]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.metrics import accuracy_score, classification_report
from imblearn.over_sampling import SMOTE

# Create a SMOTE object
smote = SMOTE(random_state=42)

# Apply SMOTE to the training set
X_train_resampled, y_train_resampled = smote.fit_resample(X_train, y_train)

# Create and train a GradientBoostingClassifier on the resampled training set
gb_classifier = GradientBoostingClassifier(n_estimators=100, random_state=42)
gb_classifier.fit(X_train_resampled, y_train_resampled)

# Make predictions on the test set
y_pred = gb_classifier.predict(X_test)

# Evaluate the model
accuracy = accuracy_score(y_test, y_pred)
print(f'Accuracy: {accuracy:.2f}')

# Print classification report
print('\nClassification Report:')
print(classification_report(y_test, y_pred))

Accuracy: 0.99

Classification Report:
              precision    recall  f1-score   support

    T3 toxic       0.00      0.00      0.00         2
      goitre       1.00      1.00      1.00         4
hyperthyroid       0.83      1.00      0.91        10
    negative       1.00      1.00      1.00       706

    accuracy                           0.99       722
   macro avg       0.71      0.75      0.73       722
weighted avg       0.99      0.99      0.99       722

